### ISSUE TITLE
Dockerizing maxwell

### ISSUE THREAD

**Issue by PhilippePitzClairoux** [zendesk/maxwell#1293]
Greetings!

I'm currently using Maxwell in a Proof of Concept and i've been having issues with docker. I have a docker-compose file wich includes mariadb, maxwell and rabbitmq. Even tho I tell docker to wait for mariadb and rabbitmq to setup before launching maxwell, maxwell keeps on launching prematurely and throws the following exception : 

> **java.lang.RuntimeException: java.net.ConnectException: Connection refused (Connection refused)**

What I currently do in order to mitigate this issue is pretty simple, I add a 15 seconds delay before launchging maxwell. It's not an elegant solution, maxwell still has a 33% crash rate.

Would it be possible to add some kind of loop that tries to reconnect when this kind of error happens? I think it could be convenient to add a "--retry" argument.

I think this would make the integration of maxwell easier and overall better!

**Comment by osheroff** [comment]
Sure, a PR to do this would be fine

**Comment by acrois** [comment]
The problem with "--retry" is that the name of the flag implies that whenever a connection dies it will attempt to reconnect. If that's the functionality you propose, a more verbose flag should be named.

Using interval of time vs retries is overall more consistent (note: failure can take longer than specified time interval, but not less) because the attempt to connect could be made and fail in a varying amount of time whereas waiting for a specific interval of time to pass before allowing the application to fail will be more appropriate for waiting for a database to start up.


In the meantime, for those interested in solutions to these sorts of problems:
https://docs.docker.com/compose/startup-order/

TL;DR: Write a script that will be called first that waits for mysql/mariadb to be ready, only running Maxwell when the database becomes available.

**Comment by PhilippePitzClairoux** [comment]
@acrois 

> The problem with "--retry" is that the name of the flag implies that whenever a connection dies it will attempt to reconnect.

I agree, would "--persistent-startup-connection" be better in your opinion?
Note: it could be an interesting feature to add connection persistence through the whole program instead of only at the begining.

> TL;DR: Write a script that will be called first that waits for mysql/mariadb to be ready, only running Maxwell when the database becomes available.

I could but I think changing the code would make it easier for me.. And others, who will undoubtedly run into the same issue. Isn't this the point of opensource software anyway? :stuck_out_tongue_closed_eyes:

I just want to clarify what i'm trying to do because I don't think I explained it very well. The best way to fix this (I think), is to add a simple loop that checks if a connection is ready (during the startup). If it's ready, the programs flow continues. But if it isn't then we wait 5 seconds before checking again.


Here's my docker-compose as a reference :
```yaml
version: "3"

services:

  rabbitmq:
    build: rabbitmq
    ports:
      - "5671:5671"
      - "5672:5672"
      - "15672:15672"

  mariadb:
    ports:
      - "3306:3306"
    environment:
      - MYSQL_ROOT_PASSWORD=bob123
    build: mariadb
    depends_on:
      - rabbitmq

  maxwell:
    environment:
      - MYSQL_USERNAME=root
      - MYSQL_PASSWORD=bob123
      - MYSQL_HOST=mariadb
      - MAXWELL_PRODUCER=rabbitmq
      - MAXWELL_OPTIONS=--config /app/config/config.properties --log_level ERROR --ignore_producer_error
    depends_on:
      - mariadb
      - rabbitmq
    build: maxwell
```

**Comment by DPGrev** [comment]
Instead of adding this functionality to maxwell. Wouldnt it be easier to have a script in your docker compose setup that waits until the mysql instance is accepting connections?

**Comment by 526654341** [comment]
I have rabbitmq connected during configuration，But the data is not pushed by mq，mq and maxwell all Normal start
